In [2]:
import sys
import pandas as pd

In [3]:
kickstarter_2018 = pd.read_csv('data\ks-projects-201801.csv')

In [4]:
pd.DataFrame([["ID", "No description provided", "Numeric"],
["name", "No description provided", "String"],
["category", "No description provided", "String"],
["main_category", "No description provided", "String"],
["currency", "No description provided", "String"],
["deadline", "No description provided", "DateTime"],
["goal", "Goal amount in project currency", "Numeric"],
["launched", "No description provided", "DateTime"],
["pledged", "Pledged amount in the project currency", "Numeric"],
["state", "No description provided", "String"],
["backers", "No description provided", "Numeric"],
["country", "No description provided", "String"],
["usd pledged", "Pledged amount in USD (conversion made by KS)", "Numeric"],
["usd_pledged_real", "Pledged amount in USD (conversion made by fixer.io api)", "Numeric"],
["usd_goal_real", "Goal amount in USD", "Numeric"]], columns=["Field name","Field description", "Type"])

Field name                                  Field description  \
0                 ID                            No description provided   
1               name                            No description provided   
2           category                            No description provided   
3      main_category                            No description provided   
4           currency                            No description provided   
5           deadline                            No description provided   
6               goal                    Goal amount in project currency   
7           launched                            No description provided   
8            pledged             Pledged amount in the project currency   
9              state                            No description provided   
10           backers                            No description provided   
11           country                            No description provided   
12       usd pledged      Pledged amount in USD (conversion made by KS)   
13  usd_pledged_real  Pledged amount in USD (conversion made by fixe...   
14     usd_goal_real                                 Goal amount in USD   

        Type  
0    Numeric  
1     String  
2     String  
3     String  
4     String  
5   DateTime  
6    Numeric  
7   DateTime  
8    Numeric  
9     String  
10   Numeric  
11    String  
12   Numeric  
13   Numeric  
14   Numeric

In [5]:
kickstarter_2018 = pd.read_csv('data\ks-projects-201801.csv', parse_dates=['deadline', 'launched'], index_col=['ID'])

### Ejercicio 1:
#### Armar una tabla con todos los proyectos con nombres duplicados, ordenados para revisar agrupados.

In [6]:
kickstarter_2018[kickstarter_2018.name.duplicated()].sort_values(by='name', ascending=False)

name  \
ID                                                              
329580179                                      xxx (Canceled)   
58345768                                         x (Canceled)   
983316237                                        x (Canceled)   
1955983697                                       x (Canceled)   
234114061                                        x (Canceled)   
774948469                                                   x   
1863478885                    weSTAND: A Stand With a Mission   
167364797                                     test (Canceled)   
1404511608                                    test (Canceled)   
216839969                                            smartLED   
576148925                                         sFTP Client   
1556035715               returning at a later date (Canceled)   
1586289411               returning at a later date (Canceled)   
391083153                returning at a later date (Canceled)   
145035452                                  project (Canceled)   
255984219                                    noparkticket.com   
1518739213                                    none (Canceled)   
421156909                                       na (Canceled)   
408045370                                      n/a (Canceled)   
513512725                                      n/a (Canceled)   
634918192                                      n/a (Canceled)   
357248791                                               mooze   
861855055                                    making egg salad   
1217993841                                              iHelp   
457581924                                               iHelp   
241601225                                                  hi   
331358040                           hand made furniture/table   
374801656                           dapper bunny: enamel pins   
563902010                  dahussen – drinnen draußen träumen   
371172741                              coming soon (Canceled)   
...                                                       ...   
741444479                                       2015 Calendar   
1535035936           1ST ANNUAL BLACK PRIDE CONCERT - TORONTO   
96794570                               12 Realms: Dungeonland   
2062423380        12 Days of Chocolate - Xmas Advent Calendar   
2061916227   12 Animal Tattoo Kits: Personalize Cool Tattoos!   
432579213                             112 Years of Upton Park   
279477490                                   1000 Paper Cranes   
552710479                                   100 Little Loaves   
470458558                                        10,000 Hours   
2004919333                                  10 Days (for now)   
1522808984                               10 Commandments Coin   
1467581633                                             1 in 3   
1889457470  $50 for a 1k word short story on a topic of yo...   
234863316                                             #Elixir   
465144515                             "What's in your spine?"   
1606189204           "The Book of Sam" Book One Graphic Novel   
2136175444              "RunRainbow" Feature Film Sizzle Reel   
516801004                                       "Reflections"   
306461885   "Pulse"- a new album from "Blind Focus" to sup...   
996180421                                       "On The Road"   
1656736114                                              "ONE"   
1637334549  "Cut From the Illest Cloth" - The Brand New Fu...   
870322805                                 "Color" Music Video   
866225086                          "City of Mercy" (Canceled)   
422509694       "American Sports Stories" - An Athletic Quest   
713417995                                     "A Fresh Start"   
816998285   "...The Last shall be first..." LODB Lifestyle...   
634871725                                                 NaN   
648853978                                                 NaN   
796533179                  

In [7]:
from hashlib import md5

In [8]:
def hashit(val):
    if isinstance(val, float):
        return str(val)
    return md5(val.encode('utf-8')).hexdigest()

### Ejercicio 2:
#### Verificar que los proyectos que tienen nombres duplicados también tienen el hash de nombre duplicado

Para verificar vamos a agregar una nueva columna 'hash_name' al dataset y guardar en ella el hash correspondiente a cada valor de 'name'. Luego vamos a contar cuantos duplicados hay para 'name' y contrastar con la cantidad de duplicados para 'hash_name'. Deberían ser iguales.

In [9]:
kickstarter_2018['hash_name'] = kickstarter_2018['name'].apply(hashit)
kickstarter_2018[kickstarter_2018.name.duplicated()].shape

(2896, 15)

In [10]:
kickstarter_2018[kickstarter_2018.hash_name.duplicated()].shape

(2896, 15)

Observamos que ambas cantidades son iguales. Con esto comprobamos que los nombres duplicados tienen el hash de nombre duplicado.

#### 1.4 Nunca modificar los datos crudos u originales

In [11]:
kickstarter_2018.to_csv('data\ks-projects-201801-for-pandas.csv')

In [12]:
kickstarter_2018 = pd.read_csv("data\ks-projects-201801-for-pandas.csv", index_col='ID', parse_dates=['deadline', 'launched'])

In [13]:
kickstarter_2018.describe(include='all')

name        category main_category currency  \
count                     378657          378661        378661   378661   
unique                    375764             159            15       14   
top     New EP/Music Development  Product Design  Film & Video      USD   
freq                          41           22314         63585   295365   
first                        NaN             NaN           NaN      NaN   
last                         NaN             NaN           NaN      NaN   
mean                         NaN             NaN           NaN      NaN   
std                          NaN             NaN           NaN      NaN   
min                          NaN             NaN           NaN      NaN   
25%                          NaN             NaN           NaN      NaN   
50%                          NaN             NaN           NaN      NaN   
75%                          NaN             NaN           NaN      NaN   
max                          NaN             NaN           NaN      NaN   

                   deadline          goal             launched       pledged  \
count                378661  3.786610e+05               378661  3.786610e+05   
unique                 3164           NaN               378089           NaN   
top     2014-08-08 00:00:00           NaN  1970-01-01 01:00:00           NaN   
freq                    705           NaN                    7           NaN   
first   2009-05-03 00:00:00           NaN  1970-01-01 01:00:00           NaN   
last    2018-03-03 00:00:00           NaN  2018-01-02 15:02:31           NaN   
mean                    NaN  4.908079e+04                  NaN  9.682979e+03   
std                     NaN  1.183391e+06                  NaN  9.563601e+04   
min                     NaN  1.000000e-02                  NaN  0.000000e+00   
25%                     NaN  2.000000e+03                  NaN  3.000000e+01   
50%                     NaN  5.200000e+03                  NaN  6.200000e+02   
75%                     NaN  1.600000e+04                  NaN  4.076000e+03   
max                     NaN  1.000000e+08                  NaN  2.033899e+07   

         state        backers country   usd pledged  usd_pledged_real  \
count   378661  378661.000000  378661  3.748640e+05      3.786610e+05   
unique       6            NaN      23           NaN               NaN   
top     failed            NaN      US           NaN               NaN   
freq    197719            NaN  292627           NaN               NaN   
first      NaN            NaN     NaN           NaN               NaN   
last       NaN            NaN     NaN           NaN               NaN   
mean       NaN     105.617476     NaN  7.036729e+03      9.058924e+03   
std        NaN     907.185035     NaN  7.863975e+04      9.097334e+04   
min        NaN       0.000000     NaN  0.000000e+00      0.000000e+00   
25%        NaN       2.000000     NaN  1.698000e+01      3.100000e+01   
50%        NaN      12.000000     NaN  3.947200e+02      6.243300e+02   
75%        NaN      56.000000     NaN  3.034090e+03      4.050000e+03   
max        NaN  219382.000000     NaN  2.033899e+07      2.033899e+07   

        usd_goal_real                         hash_name  
count    3.786610e+05                            378657  
unique            NaN                            375764  
top               NaN  2cdff1887ad0f0d341ca3f08841de3a0  
freq              NaN                                41  
first             NaN                               NaN  
last              NaN                               NaN  
mean     4.545440e+04                               NaN  
std      1.152950e+06                               NaN  
min      1.000000e-02                               NaN  
25%      2.000000e+03                               NaN  
50%      5.500000e+03                               NaN  
75%      1.550000e+04                               NaN  
max      1.663614e+08                               NaN

In [14]:
import chardet
import string
string.ascii_letters + string.digits

'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'

In [15]:
# look at the first ten thousand bytes to guess the character encoding
with open("data\ks-projects-201612.csv", 'rb') as rawdata:
    result = chardet.detect(rawdata.read(10000))
    
# check what the character encoding might be
result

{'encoding': 'Windows-1252', 'confidence': 0.73, 'language': ''}

In [16]:
# read in the file with te encoding detect by chardet
kickstarter_2016 = pd.read_csv("data\ks-projects-201612.csv", encoding = 'Windows-1252')

C:\Program Files (x86)\Microsoft Visual Studio\Shared\Anaconda3_64\lib\site-packages\IPython\core\interactiveshell.py:2785: DtypeWarning: Columns (13,14,15) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [17]:
kickstarter_2016.head()

ID                                               name   \
0  1000002330                    The Songs of Adelaide & Abullah   
1  1000004038                                     Where is Hank?   
2  1000007540  ToshiCapital Rekordz Needs Help to Complete Album   
3  1000011046  Community Film Project: The Art of Neighborhoo...   
4  1000014025                               Monarch Espresso Bar   

        category  main_category  currency             deadline   goal   \
0          Poetry     Publishing       GBP  2015-10-09 11:36:00   1000   
1  Narrative Film   Film & Video       USD  2013-02-26 00:20:50  45000   
2           Music          Music       USD  2012-04-16 04:24:11   5000   
3    Film & Video   Film & Video       USD  2015-08-29 01:00:00  19500   
4     Restaurants           Food       USD  2016-04-01 13:38:27  50000   

             launched  pledged       state  backers  country  usd pledged   \
0  2015-08-11 12:12:28        0      failed        0       GB            0   
1  2013-01-12 00:20:50      220      failed        3       US          220   
2  2012-03-17 03:24:11        1      failed        1       US            1   
3  2015-07-04 08:35:03     1283    canceled       14       US         1283   
4  2016-02-26 13:38:27    52375  successful      224       US        52375   

  Unnamed: 13 Unnamed: 14 Unnamed: 15  Unnamed: 16  
0         NaN         NaN         NaN          NaN  
1         NaN         NaN         NaN          NaN  
2         NaN         NaN         NaN          NaN  
3         NaN         NaN         NaN          NaN  
4         NaN         NaN         NaN          NaN

In [18]:
kickstarter_2018.columns = kickstarter_2018.columns.str.replace(' ', '_')
kickstarter_2018.head()

name        category  \
ID                                                                              
1000002330                    The Songs of Adelaide & Abullah          Poetry   
1000003930      Greeting From Earth: ZGAC Arts Capsule For ET  Narrative Film   
1000004038                                     Where is Hank?  Narrative Film   
1000007540  ToshiCapital Rekordz Needs Help to Complete Album           Music   
1000011046  Community Film Project: The Art of Neighborhoo...    Film & Video   

           main_category currency   deadline     goal            launched  \
ID                                                                          
1000002330    Publishing      GBP 2015-10-09   1000.0 2015-08-11 12:12:28   
1000003930  Film & Video      USD 2017-11-01  30000.0 2017-09-02 04:43:57   
1000004038  Film & Video      USD 2013-02-26  45000.0 2013-01-12 00:20:50   
1000007540         Music      USD 2012-04-16   5000.0 2012-03-17 03:24:11   
1000011046  Film & Video      USD 2015-08-29  19500.0 2015-07-04 08:35:03   

            pledged     state  backers country  usd_pledged  usd_pledged_real  \
ID                                                                              
1000002330      0.0    failed        0      GB          0.0               0.0   
1000003930   2421.0    failed       15      US        100.0            2421.0   
1000004038    220.0    failed        3      US        220.0             220.0   
1000007540      1.0    failed        1      US          1.0               1.0   
1000011046   1283.0  canceled       14      US       1283.0            1283.0   

            usd_goal_real                         hash_name  
ID                                                           
1000002330        1533.95  a6828ae8a2eca25f0dd7035efc0af0a0  
1000003930       30000.00  81609b3bdc0b96f429672d69702f2524  
1000004038       45000.00  c12f5c3bace2f0213cdb2679a265dca0  
1000007540        5000.00  4dbdcf09c86bbf5683ec39bc57b77f81  
1000011046       19500.00  9c01404a2ef702811c2088ce139042ad

### Ejercicio 3
#### Comparar la cantidad de nombres raros en kickstarter_2018 con la que obtenemos al cargar 'ks-projects-2018.csv' con enconding iso-8859-14.

In [19]:
import re
# Método que valida si un texto contiene caracteres raros
def HasStrangeCharaters(paragraph):
    flagIs = False
    words = paragraph.split(' ')
    for word in words:
        if not re.match(r'^(\w+)$', word):
            flagIs = True
    return flagIs

kickstarter_2018 = kickstarter_2018[kickstarter_2018['name'].notnull()]
kickstarter_2018[kickstarter_2018['name'].apply(HasStrangeCharaters)].shape

(241055, 15)

In [20]:
kickstarter_2018_iso_8859 = pd.read_csv("data\ks-projects-201801.csv", encoding = 'iso-8859-14')
kickstarter_2018_iso_8859 = kickstarter_2018_iso_8859[kickstarter_2018_iso_8859['name'].notnull()]
kickstarter_2018_iso_8859[kickstarter_2018_iso_8859['name'].apply(HasStrangeCharaters)].shape

(241601, 15)

Vemos que la cantidad de nombres con caracteres raros al cargar kickstarter con encoding 
iso-8859-14 es mayor a kickstarter_2018.
- kickstarter_2018 = 241055
- kickstarter_2018_iso_8859 = 241601

#### 2.2 Tratar valores faltantes

In [21]:
# get th number of missing data points per columns
missing_values_count = kickstarter_2018.isnull().sum()
# look at the # of missing points in the firts ten columns
missing_values_count[missing_values_count > 0]

usd_pledged    3797
dtype: int64

In [22]:
len(kickstarter_2018.dropna())/len(kickstarter_2018)

0.9899724552827492

In [23]:
len(kickstarter_2018.dropna(subset=['name'])) / len(kickstarter_2018)

1.0

In [24]:
kickstarter_2018[kickstarter_2018.name.isnull()]

Empty DataFrame
Columns: [name, category, main_category, currency, deadline, goal, launched, pledged, state, backers, country, usd_pledged, usd_pledged_real, usd_goal_real, hash_name]
Index: []

In [25]:
len(kickstarter_2018[kickstarter_2018.state == 'suspended'].dropna(subset=['name'])) /\
len(kickstarter_2018[kickstarter_2018.state == 'suspended'])

1.0

In [26]:
kickstarter_2018 = kickstarter_2018.dropna(subset=['name'])

In [27]:
missing_values_count = kickstarter_2018.isnull().sum()
missing_values_count[missing_values_count > 0]

usd_pledged    3797
dtype: int64

In [28]:
kickstarter_2018[kickstarter_2018.usd_pledged.isnull()]

name      category  \
ID                                                                            
1000694855                      STREETFIGHTERZ WHEELIE MURICA  Film & Video   
100149523                         Duncan Woods - Chameleon EP         Music   
1003023003          The Making of Ashley Kelley's Debut Album         Music   
1003130892                       Butter Side Down Debut Album         Music   
1003629045                            Chase Goehring debut EP         Music   
1004013077           Spencer Capier Instrumental Project 2015         Music   
1004126342  LUKAS LIGETI'S 50TH BIRTHDAY FESTIVAL: ORIGINA...         Music   
1004255433      The Battle For Breukelen: A Neighborhood Epic  Film & Video   
1004298993  "Tamboura Plays Violin" - a collection of Pop ...         Music   
1004402863        Nightingale Noel - An A Cappella Holiday CD         Music   
1005185256                             Local Music Connection         Music   
100522240   DO NOT DUPLICATE - Selected by 2015 Devon Film...  Film & Video   
1005653464                                       SA-4 Studios         Music   
1006327667                           Letters to the Wild - EP         Music   
1008671527                          The Fitness Mindset   :-)    Publishing   
1008675685               Inspired & the Sleep EP - Eyelid Kid         Music   
1008815806  Into Winter: An Instrumental Christmas Experiment         Music   
1009154868       Help advance scientific guitar & bass design         Music   
1009312309                                        Mr. Pickles    Publishing   
1009774693                Universal Peace | Universal Justice         Music   
1010217550                               Solarboot-Havel-Tour         Music   
1010694326                                          Surrender         Music   
1011508551                               EBENEZER Solo Record         Music   
1011952280                         NHSOB Summer Festival 2015         Music   
1012058337                                        Degobah Max         Music   
1012414277    WE ARE WHO WE ARE BUT WHATEVER WE WANT TO BE!!!         Music   
1012704793                               Life In The Bus Lane  Film & Video   
1012744036          An Oratorio for our Time - Last Stop Cafe         Music   
1012807897                                  Juliet Remembered  Film & Video   
1013389538  Make admission to the London Music Awards cere...         Music   
...                                                       ...           ...   
982862687                                     Bydlo - A Novel    Publishing   
982919439                                               Faith  Film & Video   
983728691                    The Octopodes' 5th Studio Album!         Music   
985231591                                            TWO FISH  Film & Video   
986253191   DEVELOP MAGAZINE FOR THE ARTS, MUSIC, AND DESI...    Publishing   
986471034      Help Revival finish their new album, Guidance.         Music   
98744618                           Vinyl for Dancing Pigeons!         Music   
988763792     Maud Sings Maud: A Musical Betsy-Tacy Companion         Music   
989218116                    Keep on Movin' - The Debut Album         Music   
989499289                                           Sojourner         Music   
989874415                         Love and Soul through Sound         Music   
991409513                           Chris Copeland's First EP         Music   
991483230   Support Vocal Synergy's Second Semester and Te...         Music   
991820324                                   The Blend CD 2015         Music   
992433603                                   Jonny Come Lately  Film & Video   
994080962                                     Maximum Coppage  Film & Video   
994621918                                         Star Sailor  Film & Video   
994911312                     "Songwriter assistance" service         Music   
994933706                   

In [29]:
kickstarter_2018[kickstarter_2018.usd_pledged.isnull()].describe(include='all')

name category main_category currency  \
count                       3797     3797          3797     3797   
unique                      3764       28            12       10   
top     New EP/Music Development    Music         Music      USD   
freq                          28     2387          2388     2738   
first                        NaN      NaN           NaN      NaN   
last                         NaN      NaN           NaN      NaN   
mean                         NaN      NaN           NaN      NaN   
std                          NaN      NaN           NaN      NaN   
min                          NaN      NaN           NaN      NaN   
25%                          NaN      NaN           NaN      NaN   
50%                          NaN      NaN           NaN      NaN   
75%                          NaN      NaN           NaN      NaN   
max                          NaN      NaN           NaN      NaN   

                   deadline          goal             launched       pledged  \
count                  3797  3.797000e+03                 3797  3.797000e+03   
unique                  722           NaN                 3797           NaN   
top     2015-05-01 00:00:00           NaN  2015-10-27 17:00:11           NaN   
freq                     26           NaN                    1           NaN   
first   2012-04-24 00:00:00           NaN  2012-03-14 19:05:44           NaN   
last    2016-07-07 00:00:00           NaN  2016-05-21 20:45:12           NaN   
mean                    NaN  5.413353e+03                  NaN  3.023265e+03   
std                     NaN  3.746624e+04                  NaN  4.491941e+04   
min                     NaN  1.080000e+02                  NaN  0.000000e+00   
25%                     NaN  2.000000e+03                  NaN  5.000000e+00   
50%                     NaN  3.500000e+03                  NaN  6.400000e+02   
75%                     NaN  5.000000e+03                  NaN  3.845000e+03   
max                     NaN  1.877000e+06                  NaN  2.762572e+06   

            state  backers country  usd_pledged  usd_pledged_real  \
count        3797   3797.0    3797          0.0      3.797000e+03   
unique          6      NaN       1          NaN               NaN   
top     undefined      NaN    N,0"          NaN               NaN   
freq         3562      NaN    3797          NaN               NaN   
first         NaN      NaN     NaN          NaN               NaN   
last          NaN      NaN     NaN          NaN               NaN   
mean          NaN      0.0     NaN          NaN      2.949851e+03   
std           NaN      0.0     NaN          NaN      4.488172e+04   
min           NaN      0.0     NaN          NaN      0.000000e+00   
25%           NaN      0.0     NaN          NaN      5.000000e+00   
50%           NaN      0.0     NaN          NaN      6.160000e+02   
75%           NaN      0.0     NaN          NaN      3.932230e+03   
max           NaN      0.0     NaN          NaN      2.762572e+06   

        usd_goal_real                         hash_name  
count    3.797000e+03                              3797  
unique            NaN                              3764  
top               NaN  2cdff1887ad0f0d341ca3f08841de3a0  
freq              NaN                                28  
first             NaN                               NaN  
last              NaN                               NaN  
mean     5.097306e+03                               NaN  
std      3.725879e+04                               NaN  
min      8.528000e+01                               NaN  
25%      2.250000e+03                               NaN  
50%      3.583050e+03                               NaN  
75%      5.208090e+03                               NaN  
max      1.877000e+06                               NaN

In [30]:
kickstarter_2018[(kickstarter_2018.country == 'N,0') & ~(kickstarter_2018.usd_pledged.isnull())].head()

Empty DataFrame
Columns: [name, category, main_category, currency, deadline, goal, launched, pledged, state, backers, country, usd_pledged, usd_pledged_real, usd_goal_real, hash_name]
Index: []

In [31]:
kickstarter_2018 = kickstarter_2018.fillna(0)

In [32]:
missing_values_count = kickstarter_2018.isnull().sum()
missing_values_count[missing_values_count > 0]

Series([], dtype: int64)

In [33]:
kickstarter_2018.shape

(378657, 15)

#### 2.3 Codificar variables

In [34]:
from sklearn import preprocessing

In [35]:
kickstarter_2018.describe(include=['O'])

name        category main_category currency  \
count                     378657          378657        378657   378657   
unique                    375764             159            15       14   
top     New EP/Music Development  Product Design  Film & Video      USD   
freq                          41           22313         63584   295362   

         state country                         hash_name  
count   378657  378657                            378657  
unique       6      23                            375764  
top     failed      US  2cdff1887ad0f0d341ca3f08841de3a0  
freq    197716  292624                                41

In [36]:
column='main_category'
# Create a label (category) encoder object
le = preprocessing.LabelEncoder()
# Fit the encoder to the pandas column
le.fit(kickstarter_2018[column])

LabelEncoder()

In [37]:
# View encoder mapping 
dict(zip(le.classes_,le.transform(le.classes_)))

{'Art': 0,
 'Comics': 1,
 'Crafts': 2,
 'Dance': 3,
 'Design': 4,
 'Fashion': 5,
 'Film & Video': 6,
 'Food': 7,
 'Games': 8,
 'Journalism': 9,
 'Music': 10,
 'Photography': 11,
 'Publishing': 12,
 'Technology': 13,
 'Theater': 14}

In [38]:
# Apply the fitted encoder to the pandas column
kickstarter_2018[column] = le.transform(kickstarter_2018[column])
kickstarter_2018.head()

name        category  \
ID                                                                              
1000002330                    The Songs of Adelaide & Abullah          Poetry   
1000003930      Greeting From Earth: ZGAC Arts Capsule For ET  Narrative Film   
1000004038                                     Where is Hank?  Narrative Film   
1000007540  ToshiCapital Rekordz Needs Help to Complete Album           Music   
1000011046  Community Film Project: The Art of Neighborhoo...    Film & Video   

            main_category currency   deadline     goal            launched  \
ID                                                                           
1000002330             12      GBP 2015-10-09   1000.0 2015-08-11 12:12:28   
1000003930              6      USD 2017-11-01  30000.0 2017-09-02 04:43:57   
1000004038              6      USD 2013-02-26  45000.0 2013-01-12 00:20:50   
1000007540             10      USD 2012-04-16   5000.0 2012-03-17 03:24:11   
1000011046              6      USD 2015-08-29  19500.0 2015-07-04 08:35:03   

            pledged     state  backers country  usd_pledged  usd_pledged_real  \
ID                                                                              
1000002330      0.0    failed        0      GB          0.0               0.0   
1000003930   2421.0    failed       15      US        100.0            2421.0   
1000004038    220.0    failed        3      US        220.0             220.0   
1000007540      1.0    failed        1      US          1.0               1.0   
1000011046   1283.0  canceled       14      US       1283.0            1283.0   

            usd_goal_real                         hash_name  
ID                                                           
1000002330        1533.95  a6828ae8a2eca25f0dd7035efc0af0a0  
1000003930       30000.00  81609b3bdc0b96f429672d69702f2524  
1000004038       45000.00  c12f5c3bace2f0213cdb2679a265dca0  
1000007540        5000.00  4dbdcf09c86bbf5683ec39bc57b77f81  
1000011046       19500.00  9c01404a2ef702811c2088ce139042ad

In [39]:
# Reversing encoding
kickstarter_2018[column] = le.inverse_transform(kickstarter_2018[column])
kickstarter_2018.head()

C:\Program Files (x86)\Microsoft Visual Studio\Shared\Anaconda3_64\lib\site-packages\sklearn\preprocessing\label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:


name        category  \
ID                                                                              
1000002330                    The Songs of Adelaide & Abullah          Poetry   
1000003930      Greeting From Earth: ZGAC Arts Capsule For ET  Narrative Film   
1000004038                                     Where is Hank?  Narrative Film   
1000007540  ToshiCapital Rekordz Needs Help to Complete Album           Music   
1000011046  Community Film Project: The Art of Neighborhoo...    Film & Video   

           main_category currency   deadline     goal            launched  \
ID                                                                          
1000002330    Publishing      GBP 2015-10-09   1000.0 2015-08-11 12:12:28   
1000003930  Film & Video      USD 2017-11-01  30000.0 2017-09-02 04:43:57   
1000004038  Film & Video      USD 2013-02-26  45000.0 2013-01-12 00:20:50   
1000007540         Music      USD 2012-04-16   5000.0 2012-03-17 03:24:11   
1000011046  Film & Video      USD 2015-08-29  19500.0 2015-07-04 08:35:03   

            pledged     state  backers country  usd_pledged  usd_pledged_real  \
ID                                                                              
1000002330      0.0    failed        0      GB          0.0               0.0   
1000003930   2421.0    failed       15      US        100.0            2421.0   
1000004038    220.0    failed        3      US        220.0             220.0   
1000007540      1.0    failed        1      US          1.0               1.0   
1000011046   1283.0  canceled       14      US       1283.0            1283.0   

            usd_goal_real                         hash_name  
ID                                                           
1000002330        1533.95  a6828ae8a2eca25f0dd7035efc0af0a0  
1000003930       30000.00  81609b3bdc0b96f429672d69702f2524  
1000004038       45000.00  c12f5c3bace2f0213cdb2679a265dca0  
1000007540        5000.00  4dbdcf09c86bbf5683ec39bc57b77f81  
1000011046       19500.00  9c01404a2ef702811c2088ce139042ad

In [40]:
from sklearn.preprocessing import LabelBinarizer
other_column = 'state'
lb = LabelBinarizer()
lb_results = lb.fit_transform(kickstarter_2018[other_column])
pd.DataFrame(lb_results, columns=((other_column + '_') + pd.Series(lb.classes_))).head(10)

state_canceled  state_failed  state_live  state_successful  \
0               0             1           0                 0   
1               0             1           0                 0   
2               0             1           0                 0   
3               0             1           0                 0   
4               1             0           0                 0   
5               0             0           0                 1   
6               0             0           0                 1   
7               0             1           0                 0   
8               1             0           0                 0   
9               1             0           0                 0   

   state_suspended  state_undefined  
0                0                0  
1                0                0  
2                0                0  
3                0                0  
4                0                0  
5                0                0  
6                0                0  
7                0                0  
8                0                0  
9                0                0

### Ejercicio 4
#### Codificar currency con ambas estrategias

In [41]:
col_currency = 'currency'
le = preprocessing.LabelEncoder()
le.fit(kickstarter_2018[col_currency])

LabelEncoder()

In [42]:
dict(zip(le.classes_,le.transform(le.classes_)))

{'AUD': 0,
 'CAD': 1,
 'CHF': 2,
 'DKK': 3,
 'EUR': 4,
 'GBP': 5,
 'HKD': 6,
 'JPY': 7,
 'MXN': 8,
 'NOK': 9,
 'NZD': 10,
 'SEK': 11,
 'SGD': 12,
 'USD': 13}

In [43]:
kickstarter_2018[col_currency] = le.transform(kickstarter_2018[col_currency])
kickstarter_2018.head()

name        category  \
ID                                                                              
1000002330                    The Songs of Adelaide & Abullah          Poetry   
1000003930      Greeting From Earth: ZGAC Arts Capsule For ET  Narrative Film   
1000004038                                     Where is Hank?  Narrative Film   
1000007540  ToshiCapital Rekordz Needs Help to Complete Album           Music   
1000011046  Community Film Project: The Art of Neighborhoo...    Film & Video   

           main_category  currency   deadline     goal            launched  \
ID                                                                           
1000002330    Publishing         5 2015-10-09   1000.0 2015-08-11 12:12:28   
1000003930  Film & Video        13 2017-11-01  30000.0 2017-09-02 04:43:57   
1000004038  Film & Video        13 2013-02-26  45000.0 2013-01-12 00:20:50   
1000007540         Music        13 2012-04-16   5000.0 2012-03-17 03:24:11   
1000011046  Film & Video        13 2015-08-29  19500.0 2015-07-04 08:35:03   

            pledged     state  backers country  usd_pledged  usd_pledged_real  \
ID                                                                              
1000002330      0.0    failed        0      GB          0.0               0.0   
1000003930   2421.0    failed       15      US        100.0            2421.0   
1000004038    220.0    failed        3      US        220.0             220.0   
1000007540      1.0    failed        1      US          1.0               1.0   
1000011046   1283.0  canceled       14      US       1283.0            1283.0   

            usd_goal_real                         hash_name  
ID                                                           
1000002330        1533.95  a6828ae8a2eca25f0dd7035efc0af0a0  
1000003930       30000.00  81609b3bdc0b96f429672d69702f2524  
1000004038       45000.00  c12f5c3bace2f0213cdb2679a265dca0  
1000007540        5000.00  4dbdcf09c86bbf5683ec39bc57b77f81  
1000011046       19500.00  9c01404a2ef702811c2088ce139042ad

In [44]:
kickstarter_2018[col_currency] = le.inverse_transform(kickstarter_2018[col_currency])
kickstarter_2018.head()

C:\Program Files (x86)\Microsoft Visual Studio\Shared\Anaconda3_64\lib\site-packages\sklearn\preprocessing\label.py:151: DeprecationWarning: The truth value of an empty array is ambiguous. Returning False, but in future this will result in an error. Use `array.size > 0` to check that an array is not empty.
  if diff:


name        category  \
ID                                                                              
1000002330                    The Songs of Adelaide & Abullah          Poetry   
1000003930      Greeting From Earth: ZGAC Arts Capsule For ET  Narrative Film   
1000004038                                     Where is Hank?  Narrative Film   
1000007540  ToshiCapital Rekordz Needs Help to Complete Album           Music   
1000011046  Community Film Project: The Art of Neighborhoo...    Film & Video   

           main_category currency   deadline     goal            launched  \
ID                                                                          
1000002330    Publishing      GBP 2015-10-09   1000.0 2015-08-11 12:12:28   
1000003930  Film & Video      USD 2017-11-01  30000.0 2017-09-02 04:43:57   
1000004038  Film & Video      USD 2013-02-26  45000.0 2013-01-12 00:20:50   
1000007540         Music      USD 2012-04-16   5000.0 2012-03-17 03:24:11   
1000011046  Film & Video      USD 2015-08-29  19500.0 2015-07-04 08:35:03   

            pledged     state  backers country  usd_pledged  usd_pledged_real  \
ID                                                                              
1000002330      0.0    failed        0      GB          0.0               0.0   
1000003930   2421.0    failed       15      US        100.0            2421.0   
1000004038    220.0    failed        3      US        220.0             220.0   
1000007540      1.0    failed        1      US          1.0               1.0   
1000011046   1283.0  canceled       14      US       1283.0            1283.0   

            usd_goal_real                         hash_name  
ID                                                           
1000002330        1533.95  a6828ae8a2eca25f0dd7035efc0af0a0  
1000003930       30000.00  81609b3bdc0b96f429672d69702f2524  
1000004038       45000.00  c12f5c3bace2f0213cdb2679a265dca0  
1000007540        5000.00  4dbdcf09c86bbf5683ec39bc57b77f81  
1000011046       19500.00  9c01404a2ef702811c2088ce139042ad

In [45]:
from sklearn.preprocessing import LabelBinarizer
other_currency = 'currency'
lb = LabelBinarizer()
lb_results = lb.fit_transform(kickstarter_2018[other_currency])
pd.DataFrame(lb_results, columns=((other_currency + '_') + pd.Series(lb.classes_))).head(10)

currency_AUD  currency_CAD  currency_CHF  currency_DKK  currency_EUR  \
0             0             0             0             0             0   
1             0             0             0             0             0   
2             0             0             0             0             0   
3             0             0             0             0             0   
4             0             0             0             0             0   
5             0             0             0             0             0   
6             0             0             0             0             0   
7             0             0             0             0             0   
8             0             0             0             0             0   
9             0             0             0             0             0   

   currency_GBP  currency_HKD  currency_JPY  currency_MXN  currency_NOK  \
0             1             0             0             0             0   
1             0             0             0             0             0   
2             0             0             0             0             0   
3             0             0             0             0             0   
4             0             0             0             0             0   
5             0             0             0             0             0   
6             0             0             0             0             0   
7             0             0             0             0             0   
8             0             0             0             0             0   
9             0             0             0             0             0   

   currency_NZD  currency_SEK  currency_SGD  currency_USD  
0             0             0             0             0  
1             0             0             0             1  
2             0             0             0             1  
3             0             0             0             1  
4             0             0             0             1  
5             0             0             0             1  
6             0             0             0             1  
7             0             0             0             1  
8             0             0             0             1  
9             0             0             0             1

### Ejercicio 5
#### 1. ¿Hay proyectos exitosos que no consiguieron el objetivo? Si hay, ¿Qué porcentaje si y cuál no?
#### 2. Calcular una tabla con la cantidad de proyectos por categoría principal y estado.

In [46]:
kickstarter_2018[(kickstarter_2018.state == 'successful') & (kickstarter_2018.pledged <= kickstarter_2018.goal)]

name  \
ID                                                              
1000070642                Mike Corey's Darkness & Light Album   
1000120151            Feather Cast Furled Fly Fishing Leaders   
1000153208    Phil Younger - Slow Down (Official Music Video)   
1001000566                                    WrittenWitheArt   
1001077166                   Student Operated Fabrication Lab   
1001878319                       DANCING CLASSROOMS GRVA 2013   
1002421792             "Fragile" Stephanie Mathias' First EP!   
1002478333                Natural face, feet, and body scrubs   
1002914380                               Dos Pueblos Tecnicos   
1003155427  Philip & Brittany Burton - NEW ALBUM - Nothing...   
1003436789              A Tenor's Journey To The German Lands   
1004211205  The Holiday Experience 2014 - A Jazzy Christma...   
1004637630                   Elemental - and the burning city   
1005009188               Maui to Maine 2017 | CLAYBOURNE TOUR   
100576591            Nina Jo Smith’s PEOPLE PLACES & SINGS CD   
1006042643                              Opal Zine - Issue 03.   
1007701013                                Seneschal Medallion   
1008038898    Co-create SATKIRIN's new Kirtan CD, SACRED KISS   
1008367501                  We Believe - Single & Music Video   
1008581356                            khildLike Experience...   
100975953                     A Few Brave Men: The Chosen Few   
1009838790                             EP Production Expenses   
1010584633                                    "A Fresh Start"   
1011028494  Music Video "Eyes Of An Angel" By Brendan Hopkins   
1011122747  To Tell the Truth—I want to make 20 new collag...   
1011614416                       Emerson ASB 2011 Documentary   
1012501347  Pants Exchange and Fake Up at the Tacoma Art M...   
101272268       Writing a children's book for my stepdaughter   
1012837921                            The 2018 Demon Calendar   
1014036836  Join Our Journey:  Public Art School Community...   
...                                                       ...   
98879144    BOTSO: Inspirational Film to Fund Music Schola...   
988827080                              'Cover Me'  CD Project   
988840562   Artisan Charcuterie - Skillfully Crafted Dry-C...   
989342077   Third Mask Productions Presents: Schizo Shakes...   
989710819      Spectacular Movements Documentary - finishing!   
99076947                        Poetry Enamel Pins by Pin Fix   
990976697                                              GIULIA   
991345829   Giant Floating Tilting Labyrinth Game Board @N...   
991401529           Make Hard Copy CD's for "To Tail A Heart"   
992046274            Matthew L. King NYC Residency/Internship   
992774017    Gifted Snowbuster Geocaching Event in Greenfield   
993137111   Esther: Elevated  A musical treatment of the s...   
993368794                        Waterloo Revival's First EP!   
993547820                                             Goggles   
99443264    Two Dancers, Two Cities, One Dance // You Win ...   
994445727   Interlaced: crossing oceans & hoping for accep...   
99468146                              Westbound Music Project   
99514178                                     Identify Project   
995600883                                           Eye Level   
995984065                                     AMS Productions   
996769069                                      The Last Straw   
997412316                             CoastWord Festival 2016   
997887544          "Reprobation": A Short Film by Ryan Allsop   
997898567   Bradley Shelver Contemporary Dance APAP/Reverb...   
998048649                   The Yearning Route - A Short FIlm   
998378475   [title of show] — The Chicago Storefront Premiere   
998622499   Quarter The Villain Is Making An Awesome Music...   
998724200                                   Death By Scrabble   
998849014                 A picture is worth a bunch of words   
999800713                  

In [50]:
# Porcentaje de proyectos exitosos que no consiguieron el objetivo y los que si lo consiguieron
num_records = kickstarter_2018[kickstarter_2018.state == 'successful'].shape[0]
num_records

133956

In [51]:
proy_exit_no_obj = kickstarter_2018[(kickstarter_2018.state == 'successful') & (kickstarter_2018.pledged <= kickstarter_2018.goal)].shape[0]
proy_exit_no_obj

4062

In [52]:
proy_exit_si_obj = kickstarter_2018[(kickstarter_2018.state == 'successful') & (kickstarter_2018.pledged > kickstarter_2018.goal)].shape[0]
proy_exit_si_obj

129894

In [53]:
porc_exit_si_objetivo = (proy_exit_si_obj * 100) / num_records
porc_exit_si_objetivo

96.96766102302249

In [54]:
porc_exit_no_objetivo = (proy_exit_no_obj * 100) / num_records
porc_exit_no_objetivo

3.032338976977515